File: app/src/main/java/com/loopeer/codereader/api/service/GithubService.java
package com.loopeer.codereader.api.service;

import com.loopeer.codereader.api.BaseListResponse;
import com.loopeer.codereader.model.Empty;
import com.loopeer.codereader.model.Repository;
import com.loopeer.codereader.model.Token;

import java.util.List;

import okhttp3.ResponseBody;
import retrofit2.Response;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;
import retrofit2.http.Streaming;
import retrofit2.http.Url;
import rx.Observable;

public interface GithubService {

    @Streaming
    @GET
    Observable<ResponseBody> downloadRepo(@Url String fileUrl);

    @GET("/search/repositories")
    Observable<Response<BaseListResponse<Repository>>> repositories(
            @Query("q") String keyword,
            @Query("sort") String sort,
            @Query("order") String order,
            @Query("page") int page,
            @Query("per_page") int pageSize
    );

    // Api about token
    @POST("authorizations")
    Observable<Response<Token>> createToken(@Body Token token, @Header("Authorization") String authorization);

    @GET("authorizations")
    Observable<Response<List<Token>>> listToken(@Header("Authorization") String authorization);

    @DELETE("authorizations/{id}")
    Observable<Response<Empty>> removeToken(@Header("Authorization") String authorization, @Path("id") String id);

}

File: code-reader-kt/src/main/java/com/loopeer/codereaderkt/api/service/GithubService.kt
package com.loopeer.codereaderkt.api.service

import com.loopeer.codereaderkt.api.BaseListResponse
import com.loopeer.codereaderkt.model.Empty
import com.loopeer.codereaderkt.model.Repository
import com.loopeer.codereaderkt.model.Token
import okhttp3.ResponseBody
import retrofit2.Response
import retrofit2.http.*
import rx.Observable


interface GithubService {

    @Streaming
    @GET
    fun downloadRepo(@Url fileUrl: String): Observable<ResponseBody>

    @GET("/search/repositories")
    fun repositories(
            @Query("q") keyword: String,
            @Query("sort") sort: String,
            @Query("order") order: String,
            @Query("page") page: Int,
            @Query("per_page") pageSize: Int
    ): Observable<Response<BaseListResponse<Repository>>>

    // Api about token
    @POST("authorizations")
    fun createToken(@Body token: Token, @Header("Authorization") authorization: String): Observable<Response<Token>>

    @GET("authorizations")
    fun listToken(@Header("Authorization") authorization: String): Observable<Response<List<Token>>>

    @DELETE("authorizations/{id}")
    fun removeToken(@Header("Authorization") authorization: String, @Path("id") id: String): Observable<Response<Empty>>

}
